sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs

```diff
@@ -244,6 +244,97 @@ impl TypedIntrinsicFunctionKind {
                     insert_type(TypeInfo::Boolean),
                 )
             }
+            Intrinsic::Gtf => {
+                if arguments.len() != 2 {
+                    errors.push(CompileError::IntrinsicIncorrectNumArgs {
+                        name: kind.to_string(),
+                        expected: 2,
+                        span,
+                    });
+                    return err(warnings, errors);
+                }
+
+                if type_arguments.len() != 1 {
+                    errors.push(CompileError::IntrinsicIncorrectNumTArgs {
+                        name: kind.to_string(),
+                        expected: 1,
+                        span,
+                    });
+                    return err(warnings, errors);
+                }
+
+                // Type check the first argument which is the index
+                let mut ctx = ctx
+                    .by_ref()
+                    .with_type_annotation(insert_type(TypeInfo::Unknown));
+                let index = check!(
+                    TypedExpression::type_check(ctx.by_ref(), arguments[0].clone()),
+                    return err(warnings, errors),
+                    warnings,
+                    errors
+                );
+
+                // Type check the second argument which is the tx field ID
+                let mut ctx = ctx
+                    .by_ref()
+                    .with_type_annotation(insert_type(TypeInfo::Unknown));
+                let tx_field_id = check!(
+                    TypedExpression::type_check(ctx.by_ref(), arguments[1].clone()),
+                    return err(warnings, errors),
+                    warnings,
+                    errors
+                );
+
+                // Make sure that the index argument is an unsigned integer
+                if !matches!(
+                    resolve_type(index.return_type, &index.span).unwrap(),
+                    TypeInfo::UnsignedInteger(_)
```

Comment:
I think this should be a fixed size int, probably `u64`.  Same for the field ID although that should probably be `u16` since it's a 12 bit val in the ASM.  Or maybe just make them both `u64` for simplicity.

Having them as wishy-washy any-size-integer makes documentation hard -- ideally we want to say this concretely is `__gtf<T>(index: u64, field_id: u64) -> T`.

Is that right?  `index` first, which is `RB` in the `GTF` instruction?  Should the index be `Option<u64>` to be a little more correct?  And while we're at it we could make an enum for all the field types -- or do we expect the stdlib to do that instead?

Reply:
I honestly would rather see these functions be type checked as regular functions.  We would add their signatures to the type checker before we start and then just assume they're regular functions until we get back to codegen.

Since we're enforcing the `__name` only for intrinsics they can be easily identified at that point and get special case codegen.  This might be worth looking at as a part of the global declaration refactor.

Reply:
I asked @vaivaswatha for his opinion here and he pointed out that while this makes sense, it's not really possible while we still have AST function call inlining.  Definitely worth revisiting after declaration engine, but not yet.

Reply:
Ah interesting. Alright, so kept it as is for now but change it to only expect `u64`s.

> Should the index be Option<u64> to be a little more correct?

Hmm potentially, though we'd have to do some magic here to make sure to expect a Sway `Option`.

> And while we're at it we could make an enum for all the field types

I was hoping not to bake this into the compiler as these can change. But also yeah the standard library will handle most of this.

Really the idea with this intrinsic is to keep it low level as it's only to be used by library authors. It's mean to replace an `asm` block effectively and allow the use of const-evaluatable tx field IDs (second argument) which `asm` don't allow yet.